### ISSUE TITLE
pmap correct usage

### ISSUE THREAD

**Issue by AdrienCorenflos** [google/jax#3638]
Hi,

What is the correct way of parallelising across devices for batch data that is bigger than the number of devices you have?

So far I am doing something like the following:

```python
from jax import jit, pmap, device_count
import jax.numpy as jnp

@jjit
def fun(z):
    return jnp.sum(z, -1)

__pfun = pmap(fun)

def _pfun(z, n):
    zs = jnp.stack(jnp.split(z, n))
    return __pfun(zs).flatten()

pfun = jjit(_pfun, static_argnums=1)
pfun(x, device_count())
```

This feels a bit wrong to me to virtually have to split, then join then split again (implicitly in the pmap) then join again.

Thanks,

Adrien

**Comment by AdrienCorenflos** [comment]
Also the above would obviously be better with array split: [here](https://github.com/google/jax/pull/3639)'s a draft PR for it.

**Comment by gnecula** [comment]
Hi Adrien, 

Another way to prepare the data for a pmap is to use reshape: `x.reshape((device_count(), -1) + x.shape[1:])`. This has the additional advantage that it is a virtual operation with no runtime cost.

**Comment by AdrienCorenflos** [comment]
Thanks that's much better indeed. Do you think there's a similar way around irregular splitting?

**Comment by gnecula** [comment]
Irregular means that the mapped axis does not divide evenly to the number of devices? Maybe you can pad the input before reshaping?
